crates/turbo-tasks/src/vc/read.rs

```diff
@@ -37,46 +53,53 @@ where
     T: VcValueType,
 {
     type Target = T;
+    type Repr = T;
 
     fn value_to_target_ref(value: &T) -> &Self::Target {
         value
     }
 
-    fn target_to_value(target: Self::Target) -> T {
+    fn target_to_repr(target: Self::Target) -> T {
         target
     }
+
     fn target_to_value_ref(target: &Self::Target) -> &T {
         target
     }
 }
 
 /// Representation for `#[turbo_tasks::value(transparent)]` types, where reads
 /// return a reference to the target type.
-pub struct VcTransparentRead<T, Target> {
-    _phantom: PhantomData<(T, Target)>,
+pub struct VcTransparentRead<T, Target, Repr> {
+    _phantom: PhantomData<(T, Target, Repr)>,
 }
 
-impl<T, Target> VcRead<T> for VcTransparentRead<T, Target>
+impl<T, Target, Repr> VcRead<T> for VcTransparentRead<T, Target, Repr>
 where
     T: VcValueType,
     Target: Any + Send + Sync,
+    Repr: VcValueType,
 {
     type Target = Target;
+    type Repr = Repr;
 
     fn value_to_target_ref(value: &T) -> &Self::Target {
         // Safety: the `VcValueType` implementor must guarantee that both `T` and
         // `Target` are #[repr(transparent)]. This is guaranteed by the
         // `#[turbo_tasks::value(transparent)]` macro.
         // We can't use `std::mem::transmute` here as it doesn't support generic types.
+        // See https://users.rust-lang.org/t/transmute-doesnt-work-on-generic-types/87272/9
         unsafe {
             std::mem::transmute_copy::<ManuallyDrop<&T>, &Self::Target>(&ManuallyDrop::new(value))
         }
     }
 
-    fn target_to_value(target: Self::Target) -> T {
+    fn target_to_repr(target: Self::Target) -> Self::Repr {
         // Safety: see `Self::value_to_target` above.
         unsafe {
-            std::mem::transmute_copy::<ManuallyDrop<Self::Target>, T>(&ManuallyDrop::new(target))
+            std::mem::transmute_copy::<ManuallyDrop<Self::Target>, Self::Repr>(&ManuallyDrop::new(
```

Comment:
This sounds like Repr and Target should have a Copy trait bound

Reply:
I don't think this is a requirement for `transmute_copy` to work. We just need `Self::Target` and `Self::Repr` to have the same size to avoid undefined behaviors. Since `VcValueType` is an `unsafe` trait, this is the implementor's responsibility to ensure (which is mentioned in the trait doc).